New: Suppressing Legend Items

NarasimhaShenoi opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**What is the purpose of this pull request? (Check any that's applicable)**

-   [X] Feature
-   [ ] Bug fix
-   [ ] Dependency updates
-   [ ] Documentation updates
-   [ ] Build related changes
-   [ ] Changes an existing functionality
-   [ ] Other, please explain:

**Does this introduce a breaking change?**

-   [ ] Yes
-   [X] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

**What changes did you make? (Give an overview)**
Introduced a new prop ```showElement```, which when set to false, will hide the legend when the legend is not having data. If showElement is set to true, it will display the legend and if the legend does not hold any value, then the legend will be in a disabled state.

**Add any screenshots**
When the second legend does not hold value, then the second legend will be in a disabled state.
<img width="1152" alt="Screenshot 2020-06-26 at 5 48 41 PM" src="https://user-images.githubusercontent.com/53594676/85856421-9ee5d180-b7d5-11ea-8b8d-659d165ac716.png">

When the second legend does not hold value and the showElement prop is set to false, then the second legend will be hidden.
<img width="1143" alt="Screenshot 2020-06-26 at 5 48 07 PM" src="https://user-images.githubusercontent.com/53594676/85856523-cb99e900-b7d5-11ea-80eb-439747608576.png">

sdadn commented:
Hi, does this new feature address any existing issue?

sdadn commented:
@NarasimhaShenoi Hi, does this new feature address any existing issue?

Veera81 commented:
@NarasimhaShenoi 

Thanks for adding enhancement. Is there any issue logged for enhancement?

NarasimhaShenoi commented:
@sdadn @Veera81 I don't see any issue being logged related to this.

Veera81 commented:
@NarasimhaShenoi Can you log an issue in the GitHub for visibility and also attach JIRA to the issue?

NarasimhaShenoi commented:
@Veera81  I have logged [issue](https://github.com/cerner/carbon-graphs/issues/252) in the Github and attached with the JIRA.

NarasimhaShenoi commented on docs/controls/Line.md:
```diff
@@ -206,6 +206,7 @@ Each line can have a legendOptions object in [Values](#values) level.
 | showShape     | boolean  | true    | Display Shape in the legend.                          |
 | showLine      | boolean  | false   | Display Line in the legend.                           |
 | style         | object   | {}      | Any style that can be applied. Refer [Styles](#style) |
+| showElement   | boolean  | true    | Toggle to hide legend when legend item has no data.   |
```
Updated [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/d52a2489dd24da79ed3b04f1fc11697cb0cee22b)

NarasimhaShenoi commented on src/main/js/helpers/legend.js:
```diff
@@ -44,6 +44,14 @@ const validateLegendLabel = (label) => {
+ * Hide legend when legend item has no data and showElement set to false
+ *
+ * @private
+ * @param {*} input item object processed from the input JSON
```
Updated [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/a11f7996cd85f8cee4719f9c23f2d913ad4a8aed)

NarasimhaShenoi commented on src/main/js/helpers/legend.js:
```diff
@@ -44,6 +44,14 @@ const validateLegendLabel = (label) => {
+ *
+ * @private
+ * @param {*} input item object processed from the input JSON
+ * @returns {string} returns none if legend is to be hide 
```
Updated [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/733bb46673fe197522ad7509ec84155f3807bef3)

NarasimhaShenoi commented on src/main/js/helpers/legend.js:
```diff
@@ -44,6 +44,14 @@ const validateLegendLabel = (label) => {
  */
 const getText = (text) => utils.sanitize(text);
+/**
+ * Hide legend when legend item has no data and showElement set to false
```
Updated [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/733bb46673fe197522ad7509ec84155f3807bef3)

Veera81 commented on src/main/js/helpers/legend.js:
```diff
@@ -44,6 +44,14 @@ const validateLegendLabel = (label) => {
+ * @param {object} input item object processed from the input JSON
+ * @returns {string} returns "none" if legend is to be hidden otherwise returns empty string 
+ */
+const hideLegend = (input) => input.legendOptions && input.legendOptions.showElement === false && utils.isEmptyArray(input.values)? "none": "";
```
So what would happen if user-provided data for legend and `showElement` to false?

NarasimhaShenoi replied:
if user-provided data for legend and showElement to false, then the legend will show up in the graph as usual. Legend will be hidden when there is no data.

Veera81 replied:
Have you verified the above condition?

Looks like `hideLegend` function will return "none"

- if `showElement` is false right? Then how will legend will be shown?

NarasimhaShenoi replied:
Yes, I did unscripted testing and also written unit test to cover the same.

If ```showElement``` is false and the legend had data, then ```utils.isEmptyArray(input.values)``` will be false, as a result it will return empty string.

NarasimhaShenoi commented on docs/controls/Line.md:
```diff
@@ -203,8 +203,9 @@ Each line can have a legendOptions object in [Values](#values) level.
 | Property Name | Expected | Default | Description                                           |
 | ------------- | -------- | ------- | ----------------------------------------------------- |
-| showShape     | boolean  | true    | Display Shape in the legend.                          |
+| showElement   | boolean  | true    | Toggle to hide legend when legend item has no data.   |
```
if the consumer already has the ```legendOptions``` object and doesn't have the ```showElement``` prop in their repo, then legend will display as did previously. If the legend is not holding any data, then it will be disabled by default.

Veera81 replied:
Then should it be `showElement` prop is optional?

NarasimhaShenoi replied:
Updated [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/61c9833d8058dfdce37354a2c719627d9b1c3cfb)

SharathAlur replied:
I think all the properties should be optional in the legend options!

NarasimhaShenoi replied:
All properties [moved](https://github.com/NarasimhaShenoi/carbon-graphs/commit/cd4edefe5fb917e528894c96dd60d7a29616f179) to optional in legend options

SharathAlur replied:
+1

NarasimhaShenoi commented on docs/controls/Line.md:
```diff
@@ -203,8 +203,9 @@ Each line can have a legendOptions object in [Values](#values) level.
 
 | Property Name | Expected | Default | Description                                           |
 | ------------- | -------- | ------- | ----------------------------------------------------- |
-| showShape     | boolean  | true    | Display Shape in the legend.                          |
+| showElement   | boolean  | true    | Toggle to hide legend when legend item has no data.   |
```
I have not seen the legendOptions documentation for other graphs. That's why I missed updating. Let me know where all I need to add.

NarasimhaShenoi commented on src/main/js/helpers/legend.js:
```diff
@@ -125,7 +134,7 @@ const loadLegendItem = (legendSVG, t, config, eventHandlers) => {
 const processLegendOptions = (buttonPath, input) => {
     if (input.legendOptions) {
-        if (input.legendOptions.showShape) {
+        if (input.legendOptions.showShape || input.legendOptions.showElement) {
```
Added [comment](https://github.com/NarasimhaShenoi/carbon-graphs/commit/954e6d20c65b0dec3da117f1ba8b1c0cbc0d6bc0)

NarasimhaShenoi commented on src/test/unit/controls/Line/LineLoad-spec.js:
```diff
@@ -1226,6 +1226,59 @@ describe("Line - Load", () => {
                 done();
             });
         });
+        it("hides the legend if legend has no data and user sets showElement to be false", () => {
```
Updated [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/5ace1f8eafac54f4dec704468d1992de6ae85953)

NarasimhaShenoi commented on src/test/unit/controls/Line/LineLoad-spec.js:
```diff
@@ -1226,6 +1226,59 @@ describe("Line - Load", () => {
                 done();
             });
         });
+        it("hides the legend if legend has no data and user sets showElement to be false", () => {
```
Added [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/5ace1f8eafac54f4dec704468d1992de6ae85953#diff-d1a871608782515276a74bc712afc556R1266)

Veera81 commented on docs/controls/Line.md:
```diff
@@ -203,10 +203,16 @@ Each line can have a legendOptions object in [Values](#values) level.
+#### Optional
+
+| Property Name | Expected | Default | Description                                           |
+| ------------- | -------- | ------- | ----------------------------------------------------- |
```
Do you know what other graphs will have this option? We probably update their documentation as well

NarasimhaShenoi replied:
I see legendOptions is only used by pairedResult and Line graphs. I have updated their documentation [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/a3c0b7b1e03a373d11caeeaa665926ed3bc3a526)

Veera81 replied:
Thanks for updating the documentation. Does it mean it is not required for other graphs?

NarasimhaShenoi replied:
The legend option is available only for the line and paired result. It is not required for other graphs.

SharathAlur replied:
I believe showElement should be available for all the graph types. Other properties like, showShape, showLine and style only applied to line and paired result graphs. So we can expose only showElement on all graph types.

NarasimhaShenoi replied:
Updated documentation for all graphs [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/cd4edefe5fb917e528894c96dd60d7a29616f179)

SharathAlur replied:
+1

jeremyfuksa commented:
Hi, everyone.

When asking for UX review, can you please leave a comment with the dev site example I need to be looking at? I've built this PR and the example labels don't give me an idea of what I need to be looking at.

Veera81 commented:
@NarasimhaShenoi Can you please add dev site example for your changes.

cc: @jeremyfuksa

NarasimhaShenoi commented:
Hi @jeremyfuksa, I have added in dev site examples under the "Suppress Legend" section. I added two legends( "Data Label 1", "Data Label 8") here, where the first legend has data and am suppressing the second legend. As a result, only the first legend is shown.

cc: @Veera81

NarasimhaShenoi commented:
I have added the [suppress legend items ](https://github.com/NarasimhaShenoi/carbon-graphs/commit/10369e15f4cf8d7b25488bad5c5d6774d75b4f9f), updated the [method name](https://github.com/NarasimhaShenoi/carbon-graphs/commit/4876c1cbdcf1488dec06e2964f9c0cb395e8f2f5) from hideLegend to legendDispalyStyle for more clarity, [removed showElement](https://github.com/NarasimhaShenoi/carbon-graphs/commit/e433bc110c99f3f160ab1cd59fbcda75f029dc22) from processLegendOptions method and finally [updated spelling mistake](https://github.com/NarasimhaShenoi/carbon-graphs/commit/db394e99fb34cdbfe0f2d8623f329e3d67bf914a) i gave for the dev site example i added.

NarasimhaShenoi commented:
@Veera81 @jeremyfuksa I have made the requested changes. Please review when you get chance.

Veera81 commented on docs/controls/PairedResult.md:
```diff
@@ -108,16 +108,35 @@ Refer [Graph](../core/Graph.md) `Root` for more details.
+| color         | string   | COLORS.BLACK       | Color for the data point                                                          |
+| shape         | string   | SHAPES.DARK.CIRCLE | Shape for representing the data points                                            |
+| onClick       | Function | undefined          | Any action that can be performed when clicking on the data point                  |
+| legendOptions | object   | undefined          | Option to show/hide shape and/or line. Refer [LegendOption(#legendOptions)        |
```
Refer [LegendOptions link is not working

NarasimhaShenoi replied:
Fixed [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/0c528924d29c82332690c3ec3e7d77aa180c8b9a#diff-ac5dee9f2f5653ec431590b47bc921f2R121)

Veera81 commented on docs/controls/PairedResult.md:
```diff
@@ -108,16 +108,35 @@ Refer [Graph](../core/Graph.md) `Root` for more details.
+| ------------- | -------- | ------- | ----------------------------------------------------- |
+| showLine      | boolean  | false   | Display Line in the legend.                           |
+| showShape     | boolean  | true    | Display Shape in the legend.                          |
+| style         | object   | {}      | Any style that can be applied. Refer [Styles](#style) |
```
Styles link also not working

NarasimhaShenoi replied:
Fixed [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/0c528924d29c82332690c3ec3e7d77aa180c8b9a#diff-ac5dee9f2f5653ec431590b47bc921f2R141)

sdadn commented on docs/controls/PairedResult.md:
```diff
@@ -108,16 +108,40 @@ Refer [Graph](../core/Graph.md) `Root` for more details.
+
+| Property Name | Expected | Default            | Description                                                                       |
+| ------------- | -------- | ------------------ | ----------------------------------------------------------------------------------|
+| yAxis         | string   | "y"                | Setting for using different Y based axis. For now: its either Y or Y2             |
+| regions       | object   | {}                 | Refer [Regions](#regions)                                                         |
+| label         | object   | {}                 | Display value for the data-set which the data points belong to                    |
+| color         | string   | COLORS.BLACK       | Color for the data point                                                          |
+| shape         | string   | SHAPES.DARK.CIRCLE | Shape for representing the data points                                            |
+| onClick       | Function | undefined          | Any action that can be performed when clicking on the data point                  |
+| legendOptions | object   | undefined          | Toggle to show shape, line and legend. Refer [LegendOptions](#legendOptions)      |
```
Nit: properties are not alphabetized.

NarasimhaShenoi replied:
Updated [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/1f13594bf10e008d1076eec74e3a34dd088c0384)

sdadn commented on docs/controls/Scatter.md:
```diff
@@ -140,14 +140,15 @@ Refer [Graph](../core/Graph.md) `Root` for more details.
-| onClick       | Function | undefined          | Any action that can be performed when clicking on the data point      |
+| Property Name | Expected | Default            | Description                                                                       |
+| ------------- | -------- | ------------------ | --------------------------------------------------------------------------------- |
+| yAxis         | string   | "y"                | Setting for using different Y based axis. For now: its either Y or Y2             |
+| regions       | array    | []                 | Refer [Regions](#regions)                                                         |
+| label         | object   | {}                 | Display value for the data-set which the data points belong to                    |
+| color         | string   | COLORS.BLACK       | Color for the data point                                                          |
+| shape         | string   | SHAPES.DARK.CIRCLE | Shape for representing the data points                                            |
+| onClick       | Function | undefined          | Any action that can be performed when clicking on the data point                  |
+| legendOptions | object   | undefined          | Toggle to show legend. Refer [LegendOptions](#legendOptions)                      |
```
Nit: properties are not alphabetized.

NarasimhaShenoi replied:
Updated [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/1f13594bf10e008d1076eec74e3a34dd088c0384)

sdadn commented on src/test/unit/controls/Line/LineLoad-spec.js:
```diff
@@ -1226,6 +1226,91 @@ describe("Line - Load", () => {
+            it("should hide the legend", () => {
+                const input = getInput([], false, false);
+                input.legendOptions = {
+                    showElement: false
+                }
+                graphDefault.loadContent(new Line(input));
+                const legendContainer = fetchElementByClass(
+                    lineGraphContainer,
+                    styles.legendItem
+                );
+                expect(legendContainer).not.toBeNull();
+                expect(legendContainer.tagName).toBe("LI");
+                const legendItem = document.body.querySelector(
+                    `.${styles.legendItem}`
+                );
+                expect(legendItem.getAttribute("style")).toBe("display: none; padding: 4px 8px;");
+            })
+        });
+        describe("if legend has no data and user sets showElement to be true", () => {
+            it("should show the legend", () => {
+                const input = getInput([], false, false);
+                input.legendOptions = {
+                    showElement: true,
+                }
+                graphDefault.loadContent(new Line(input));
+                const legendContainer = fetchElementByClass(
+                    lineGraphContainer,
+                    styles.legendItem
+                );
+                const legendItems = legendContainer.children;
+                expect(legendContainer).not.toBeNull();
+                expect(legendContainer.tagName).toBe("LI");
+                expect(legendItems.length).toBe(2);
+                const legendItem = document.body.querySelector(
+                    `.${styles.legendItem}`
+                );
+                expect(legendItem.getAttribute("aria-disabled")).toBe("true");
+                expect(legendItem.getAttribute("aria-current")).toBe("true");
+            })
+        });
```
Describe blocks are meant to group together similar tests. These tests can be combined into a single describe block.

```js
describe("when the legend has no data",  ()=>{
   it ("should hide the legend if showElement is false", ()=>{....})
        ......
   it ("should show the legend if showElement is true", ()=>{....})
        ......
}
```

NarasimhaShenoi replied:
Updated [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/e5eae0edb49ef09342b04de3df0500d909b9e134)

sdadn commented on src/test/unit/controls/Line/LineLoad-spec.js:
```diff
@@ -1226,6 +1226,91 @@ describe("Line - Load", () => {
+                    showElement: true,
+                }
+                graphDefault.loadContent(new Line(input));
+                const legendContainer = fetchElementByClass(
+                    lineGraphContainer,
+                    styles.legendItem
+                );
+                const legendItems = legendContainer.children;
+                expect(legendContainer).not.toBeNull();
+                expect(legendContainer.tagName).toBe("LI");
+                expect(legendItems.length).toBe(2);
+                const legendItem = document.body.querySelector(
+                    `.${styles.legendItem}`
+                );
+                expect(legendItem.getAttribute("aria-disabled")).toBe("false");
+                expect(legendItem.getAttribute("aria-current")).toBe("true");
+            })
+        });
+        describe("if legend has data and user sets showElement to be true", () => {
+            it("should show the legend", () => {
+                const input = getInput(valuesDefault, false, false);
+                input.legendOptions = {
+                    showElement: true,
+                }
+                graphDefault.loadContent(new Line(input));
+                const legendContainer = fetchElementByClass(
+                    lineGraphContainer,
+                    styles.legend
+                );
+                const legendItems = legendContainer.children;
+                expect(legendContainer).not.toBeNull();
+                expect(legendContainer.tagName).toBe("UL");
+                expect(legendItems.length).toBe(1);
+                const legendItem = document.body.querySelector(
+                    `.${styles.legendItem}`
+                );
+                expect(legendItem.getAttribute("aria-disabled")).toBe("false");
+                expect(legendItem.getAttribute("aria-current")).toBe("true");
+            });
+        });
```
These tests can also be combined:

```js
describe("when the legend has data",  ()=>{
   it ("should hide the legend if showElement is false", ()=>{....})
        ......
   it ("should show the legend if showElement is true", ()=>{....})
        ......
}
```

NarasimhaShenoi replied:
Updated [here](https://github.com/NarasimhaShenoi/carbon-graphs/commit/e5eae0edb49ef09342b04de3df0500d909b9e134)

ryanthemanuel commented:
:tada: This PR is included in version 2.8.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@cerner/carbon-graphs/v/2.8.0)
- [GitHub release](https://github.com/cerner/carbon-graphs/releases/tag/v2.8.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
